### PR TITLE
Home seeker

### PR DESCRIPTION
**Pull request by KevinZheng0701**

### CODE REVIEW

File: src/app/components/Navbar/actions.ts
Comment (MikeRomaa) [inline src/app/components/Navbar/actions.ts:1]: Was this file left in accidentally? Or is there something planned for this?

File: src/app/components/appointmentform/actions.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	start_time: Date,
+	end_time: Date,
+): Promise<Appointment | string> {
+	// Errors checking
+	if (!start_time) {
+		return "Start time required.";
+	}
+	if (!end_time) {
+		return "End time required.";
+	}
```
Comment (MikeRomaa) [inline src/app/components/appointmentform/actions.ts:23]: Don't know much much of a tall order this is (if it's too much I can just do it myself) but can we maybe use the `FormStatus` type and provide these as `fieldErrors`? There's an example of this in the `sign-in/actions.ts` file.

Main reason for this is that it allows us to show this validation data on the `TextInput` components.

File: src/app/components/appointmentform/actions.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		if (start_time < appointment.start) {
+			if (end_time > appointment.start) {
+				return "Time conflicts with another appointment.";
+			}
+		} else {
+			if (start_time < appointment.end) {
+				return "Time conflicts with another appointment.";
+			}
+		}
+	}
```
Comment (MikeRomaa) [inline src/app/components/appointmentform/actions.ts:69]: We could probably convert this to a SQL query which would be faster, but I'm fine with just leaving it as-is for the time being.

File: src/app/auth/register/page.tsx [hunk 3]
```diff
@@ -9,7 +9,6 @@ import { State, register } from "@/app/auth/register/actions";
 
 const Register: NextPage = () => {
 	const [state, formAction] = useFormState<State, FormData>(register, {});
-
```
Comment (MikeRomaa) [inline src/app/auth/register/page.tsx]: Ehhhh leave the blank line there. I like having some separation between hooks and the actual React elements.

File: src/db/index.ts
Comment (MikeRomaa) [inline src/db/index.ts:1]: Can we also revert all of the changes in this file?

File: src/db/tutortime/services.ts
Comment (MikeRomaa) [inline src/db/tutortime/services.ts:1]: Aaaaand this one too? Sorry lol, just want to make sure we aren't making this commit bigger than it needs to be.

File: src/db/homeseeker/schedule.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	start: Date;
+	end: Date;
+}
+
+/**
+ * Retrieves all schedules made from database.
+ */
+export async function getSchedules(): Promise<Schedule[]> {
+	const [res] = await pool.execute<Schedule[]>(
+		"SELECT * FROM bookings_db.hs_schedule",
```
Comment (MikeRomaa) [inline src/db/homeseeker/schedule.ts]: It's generally considered bad practice to use `SELECT *` for data-fetching queries since you are (a) hiding the columns away from anybody reading the file, and (b) relying on the schema not changing. Instead, it's best to list out all of the columns you need, even if it is all of them.

File: src/db/homeseeker/property.ts [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+	built?: number,
+): Promise<void> {
+	const params = {
+		address,
+		zipcode,
+		type,
+		price: price !== undefined ? price : null,
+		rooms: rooms !== undefined ? rooms : null,
+		area: area !== undefined ? area : null,
+		built: built !== undefined ? built : null,
```
Comment (MikeRomaa) [inline src/db/homeseeker/property.ts:174]: Is the intention with these to make the function arguments optional? Or is it so you can explicitly pass in `undefined` or `null` to clear the column?

File: src/app/user/actions.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+export async function getUserAppointments(user_id: number) {
+    const appointments = await getUpcomingAppointmentsByBroker(user_id);
+    return appointments;
+}
+
+export async function fetchUserDetails(): Promise<User | null> {
+    const user = await getCurrentUser();
+    return user;
+}
```
Comment (MikeRomaa) [inline src/app/user/actions.ts]: With all of these simple actions you should be able to just return the promise directly from the database function. For example,
```ts
return getPropertiesByUser(user_id);
```

File: src/app/homeseeker/page.tsx [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	const properties = await getProperties();
+	const user = getCurrentUser();
+	// Not signed it so just display everything
+	if (!user) {
+		return properties;
+	}
+	const filteredProperties = properties.filter((property) => {
+		return property.broker_id !== user.id
+	})
+	return filteredProperties;
```
Comment (MikeRomaa) [inline src/app/homeseeker/page.tsx]: Perhaps this should be an option? Or maybe you can show a little badge on listings that are posted by the current user? I feel like hiding *all other listings* when you're signed in is a bit weird.

File: src/app/homeseeker/display/page.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+									Start time: {new Date(schedule.start).toLocaleString()}
+								</p>
+								<p className="text-slate-600 text-sm">
+									End time: {new Date(schedule.end).toLocaleString()}
+								</p>
+							</div>
+							<Link href={{
+								pathname: '/homeseeker/makeappointment', query: {
+									schedule: schedule.id
+								}
```
Comment (MikeRomaa) [inline src/app/homeseeker/display/page.tsx]: Not sure if this is Biome's doing or not, but the formatting is a bit weird.
```suggestion
								pathname: '/homeseeker/makeappointment',
								query: { schedule: schedule.id },
```

File: src/app/homeseeker/display/actions.ts
Comment (MikeRomaa) [inline src/app/homeseeker/display/actions.ts:1]: Think we saw one of these earlier but is this intentional?

File: src/app/homeseeker/actions.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    const user = getCurrentUser();
+    // Not signed it so just display everything
+    if (!user) {
+        return properties;
+    }
+    const filteredProperties = properties.filter((property) => {
+        return property.broker_id !== user.id
+    })
+    return filteredProperties;
+}
```
Comment (MikeRomaa) [inline src/app/homeseeker/actions.ts]: Wasn't this exact code included on the `page.tsx` file?

File: src/app/globals.css
Comment (MikeRomaa) [inline src/app/globals.css:1]: Similarly to before, lets revert these changes.

File: src/app/components/scheduleform/page.tsx [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+            const output = await makeSchedule(
+                property_id,
+                new Date(startDateTime),
+                new Date(endDateTime)
+            )
+            if (typeof output === "string") {
+                setError(output);
+            } else {
+                setError(null);
+                window.location.reload();
```
Comment (MikeRomaa) [inline src/app/components/scheduleform/page.tsx]: Better to use `router.refresh()` here, since it plays nicer with the slightly weird way routing works in Next.js. See the [`useRouter` documentation](https://nextjs.org/docs/app/api-reference/functions/use-router) for more info.

File: src/app/components/scheduleform/page.tsx [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        <div className="container mx-auto py-20">
+            {property_id ? (
+                <div>
+                    <Card className="max-w-96 text-center">
+                        <h2>Upload your availability</h2>
+                        <form onSubmit={handleSubmit} className="flex flex-col justify-center items-center gap-2 mt-3">
+                            <div className="flex-1 w-full">
+                                <div className="w-full flex justify-center items-center m-2">
+                                    <label htmlFor="start">Date</label>
+                                    <input
```
Comment (MikeRomaa) [inline src/app/components/scheduleform/page.tsx]: These should be the Tremor UI components, not bog standard inputs.

**Review by MikeRomaa — CHANGES_REQUESTED** [review]
None of these comments are demands, and if you disgree or have questions by all means let's discuss them.

File: src/app/components/appointmentform/actions.ts (hunk 1 above)
Comment (KevinZheng0701) [reply src/app/components/appointmentform/actions.ts:23]: Honestly we dont need to check for that since in the page the time are set to required. The reason we didn't use the textinput along with the formstatus was something regarding: Type '"time"' is not assignable to type '"text" | "password" | "email" | "url" | undefined'.

File: src/app/components/Navbar/actions.ts
Comment (KevinZheng0701) [reply src/app/components/Navbar/actions.ts:1]: Before I tried to do something with the navbar but it didn't work as I would have wanted it, so I deleted the content inside and forgot about deleting that file.

File: src/app/components/appointmentform/actions.ts (hunk 2 above)
Comment (KevinZheng0701) [reply src/app/components/appointmentform/actions.ts:69]: Yea probably faster with SQL.

File: src/app/auth/register/page.tsx (hunk 3 above)
Comment (KevinZheng0701) [reply src/app/auth/register/page.tsx]: Of course!

File: src/db/homeseeker/schedule.ts (hunk 4 above)
Comment (KevinZheng0701) [reply src/db/homeseeker/schedule.ts]: Got it

File: src/db/homeseeker/property.ts (hunk 5 above)
Comment (KevinZheng0701) [reply src/db/homeseeker/property.ts:174]: I think the intention was for it to be optional parameter and in SQL they accept null and not undefined but in the page if someone didn't fill some part of the form it is undefined. So this basically make the SQL query have the correct type when updating/inserting.

File: src/app/user/actions.ts (hunk 6 above)
Comment (KevinZheng0701) [reply src/app/user/actions.ts]: I think it was like this so when developing i can console the output and see if it is correct. But yea once everything is done correctly, these can be simplified.

File: src/app/homeseeker/page.tsx (hunk 7 above)
Comment (KevinZheng0701) [reply src/app/homeseeker/page.tsx]: I was thinking if this is a large scale home seeker app, it doesn't make sense for user to view their own homes. They would want to see other property listed. For now they can still view their registered properties in their account page. But maybe you're right we can just add something to show that the property is owned by a certain user instead of filtering them.

File: src/app/homeseeker/display/page.tsx (hunk 8 above)
Comment (KevinZheng0701) [reply src/app/homeseeker/display/page.tsx]: Probably i didn't turn off prettier when doing this project that's probably why.

File: src/app/components/scheduleform/page.tsx (hunk 10 above)
Comment (KevinZheng0701) [reply src/app/components/scheduleform/page.tsx]: sure

File: src/app/components/scheduleform/page.tsx (hunk 11 above)
Comment (KevinZheng0701) [reply src/app/components/scheduleform/page.tsx]: I think there is a data selector for this from tremor. Might use it instead. But don't think there are timeinputs from tremor.

Comment (MikeRomaa) [reply src/app/components/scheduleform/page.tsx]: Oh hmm you're right. I'll make a component so we can leave this in for now.

**Review by MikeRomaa — APPROVED** [review]
